feat: implement blind signature command

dj8yfo opened the pull request.

mr-presto committed:
test: fix doc, add examples, add signature verify test

dj8yfo commented on src/lib.rs:
```diff
@@ -75,21 +87,19 @@ pub fn get_version() -> Result<NEARLedgerAppVersion, NEARLedgerError> {
                 return Ok(response.data().to_vec());
             } else {
-                let error_string = response
-                    .error_code()
```
`ledger-apdu-0.10.0` error code conversion table is probably from another app, its codes do not correspond to the ones in the `app-near` ledger repo

frol commented on src/lib.rs:
````diff
@@ -229,11 +230,12 @@ pub fn get_public_key(
+///     .expect("Signature is not expected to fail on deserialization");
 /// ```
 pub fn sign_transaction(
-    unsigned_transaction_borsh_serializer: BorshSerializedUnsignedTransaction,
+    target: SignTarget,
````
To be honest, from a crate user perspective, I would prefer to have a single `sign_transaction` method that would figure out when it should use blind signing or not, and maybe additionally have 3 lower-level methods:

* non-blind transaction signing (receive a serialized transaction)
* blind transaction signing (receive a serialized transaction)
* blind transaction hash signing (receive a hash)

dj8yfo commented:
discovered a marginal bug when reiterating a test with signing a normal tx, fitting in buffer
[LedgerHQ/app-near#31](https://github.com/LedgerHQ/app-near/pull/31)

mr-presto committed:
build: fix doc-test

mr-presto committed:
feat: add `OnlyBlindSigning` struct

mr-presto committed:
test: add signature verification to `blind_sign_transaction` example

mr-presto committed:
chore: add error enum variant for blind sig disabled in settings

mr-presto committed:
chore: replace `sha2` dep with `near_primitives_core::hash::CryptoHash`

frol commented on examples/blind_sign_transaction.rs:
```diff
@@ -0,0 +1,91 @@
+use slip10::BIP32Path;
+
+fn too_long_tx() -> near_primitives::transaction::Transaction {
+    let ledger_pub_key_str = "7be181e38cf773f8432a1af401f83b39f1222bad5a167875abba1baa2de477c7";
+    let receiver_pub_key_str = "dc7e34eecec3096a4a661e10932834f801149c49dba9b93322f6d9de18047f9c";
+    let ledger_pub_key_bytes = hex::decode(ledger_pub_key_str).unwrap();
```
Given that public/secret keys are encoded with base58 in NEAR, let's use base58 here as well to avoid confusion.

dj8yfo replied:
the hex representation is used for computing `signer_id`/`receiver_id` `AccountId`-s
Attempting to do:
```rust
    let account_id = AccountId::from_str("9Lab2du3zZg5MrZPJCAd1fka7nHuP9Rgo643d8C8hGiJ").unwrap();
```
results in 
```bash
called `Result::unwrap()` on an `Err` value: ParseAccountError { kind: InvalidChar, char: Some((1, 'L')) }
```
and
```rust
    let account_id = AccountId::from_str("ed25519:9Lab2du3zZg5MrZPJCAd1fka7nHuP9Rgo643d8C8hGiJ").unwrap();
```
in 
```bash
called `Result::unwrap()` on an `Err` value: ParseAccountError { kind: InvalidChar, char: Some((7, ':')) }
```
I'll add a commit which doesn't use the hex `str` for computing the pub key for transaction, but instead uses the public key queried from ledger (this should've been so initially for 2 signature examples)

dj8yfo replied:
addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/6b030733d641434b06982cb51862d2684c435a4e

frol commented on examples/blind_sign_transaction.rs:
```diff
@@ -0,0 +1,91 @@
+    let block_hash_bytes = block_hash_str.from_base58().unwrap();
+    let mut block_hash: [u8; 32] = [0; 32];
+    block_hash.copy_from_slice(&block_hash_bytes[0..32]);
+    let block_hash = near_primitives::hash::CryptoHash(block_hash);
```
Use `CryptoHash::from_str` or even `"...".parse::<CryptoHash>()`

dj8yfo replied:
addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/e80c33931f219b1363680dd2786ccfd0b07a6f4f

frol commented on examples/sign_transaction.rs:
```diff
@@ -35,18 +37,41 @@ fn main() -> Result<(), NEARLedgerError> {
+    let hash = CryptoHash::hash_bytes(bytes);
+
+    pub_key.verify(&hash.as_ref(), &signature)
+}
```
It feels like the function body is actually simpler than the signature. I would just inline it

dj8yfo replied:
addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/6c7d11440a8c1f391e4d11aa726354f7885a03c5

frol commented on examples/blind_sign_transaction.rs:
```diff
@@ -0,0 +1,91 @@
+    let hash = OnlyBlindSigning::hash_bytes(&bytes);
+    log::info!("{:<25} : {}", "hash (hex)", hex::encode(hash.0.as_ref()));
+    log::info!("{:<25} : {}", "hash", hash.0);
+    assert!(matches!(err, NEARLedgerError::BufferOverflow(err_hash) if err_hash == hash));
```
This looks like an abuse of an example. It seems like you wanted to implement a test and example at the same time, but the downside now is that the example is hard to follow due to verbosity.

dj8yfo replied:
assert of signature verification was useful for judging about correctness with minimal env (without need to attempt to send transactions over network) in this https://github.com/LedgerHQ/app-near/pull/27#discussion_r1385223992 case and any other, possibly future, case. 
Setting up fully automated tests (running in ci, not requiring an actual device) is possible, but will probably be hard/cumbersome task.
Hopefully this comment is addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/1da9710f0d472d2dd91c232ce1fc941e01b94b29

frol commented on src/lib.rs:
```diff
@@ -169,28 +193,41 @@ pub fn get_public_key(
+    };
+    match TransportNativeHID::new(&hidapi) {
+        Ok(transport) => Ok(transport),
+        // TODO: refactor this
+        // https://github.com/Zondax/ledger-rs/issues/65
+        Err(_err) => Err(NEARLedgerError::DeviceNotFound),
```
Since the issue is resolved, can we address this TODO?

dj8yfo replied:
addressed in [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/25b16bd49998b08d85aba6ce177ce453dc2c4948)

frol commented on src/lib.rs:
```diff
@@ -27,8 +41,23 @@ pub enum NEARLedgerError {
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct OnlyBlindSigning(pub CryptoHash);
```
I would use struct with a named field here to explicitly say what that CryptoHash is going to be.

dj8yfo replied:
addressed in [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/a179ffa6791569cb599f630bab163bf1129f9c4c)

dj8yfo replied:
addressed in [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/26d0f39f09c5d59f7d9a8c029041a795022d4c67)

frol commented on src/lib.rs:
```diff
@@ -27,8 +41,23 @@ pub enum NEARLedgerError {
+pub struct OnlyBlindSigning(pub CryptoHash);
+
+impl OnlyBlindSigning {
+    pub fn hash_bytes(bytes: &[u8]) -> Self {
```
I don't see any value in exposing this method to be part of the public API. Do you?

dj8yfo replied:
addressed by [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/245d912a5ae3c6dfff2af450334989f48a4e4fbd)

frol reviewed (changes requested):
Nice! I have only left some nit comments, but overall it is in a pretty good state

mr-presto committed:
chore: [address review comment] replace `NEARLedgerError::DeviceNotFound`

mr-presto committed:
chore: [address review comment] named field in `OnlyBlindSigning`

mr-presto committed:
chore: [address review comment] non-pub `hash_bytes` method

resolves https://github.com/khorolets/near-ledger-rs/pull/4#discussion_r1406815025

mr-presto committed:
chore: [address review comment] inline `verify_near` helper fn

resolves https://github.com/khorolets/near-ledger-rs/pull/4#discussion_r1406791937

mr-presto committed:
chore: [address review comments] avoid mentioning tx's public_key in `hex` format

resolves https://github.com/khorolets/near-ledger-rs/pull/4#discussion_r140678524

mr-presto committed:
chore: [address review comment] use `"...".parse::<CryptoHash>()`

resolves https://github.com/khorolets/near-ledger-rs/pull/4#discussion_r1406786986

mr-presto committed:
chore: [address review comment] improve readability of examples

resolves https://github.com/khorolets/near-ledger-rs/pull/4#discussion_r1406798758

frol commented on src/lib.rs:
```diff
@@ -16,19 +20,53 @@ const CHUNK_SIZE: usize = 128; // Chunk size to be sent to Ledger
+#[derive(Debug, PartialEq, Eq)]
+pub struct OnlyBlindSigning {
+    pub byte_array: CryptoHash,
+}
```
How about turning it into?

```suggestion
    BufferOverflow { transaction_hash: CryptoHash },
}
```

given that we only pull in near-primitives-core for CryptoHash, we may even consider having our own `type CryptoHash = [u8; 32];`

dj8yfo replied:
addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/7729ef8bc0782688d2604f54f68109cfd0eb1865

frol reviewed (approved):
I have one last question/suggestion, but overall I believe it is already in a good shape, so I approve it.

mr-presto committed:
chore: [address review comment] remove redundant `OnlyBlindSigning` type

resolves https://github.com/khorolets/near-ledger-rs/pull/4#discussion_r1408479085
